### PR TITLE
Shrink tlab to capacity

### PR DESCRIPTION
**Pull request by kdnilsen**
When a TLAB request exceeds the currently available memory within young-gen, the existing behavior is to reject the TLAB request outright. This is recognized as a failed allocation request, which triggers degenerated GC.

This change introduces code to reduce the likelihood that too-large TLAB requests will be issued, and when they are issued, it makes an effort to shrink the TLAB request in order to reduce the need for degenerated GC.

The impact is difficult to measure because this situation is fairly rare.  On one Extremem workload, the TLAB-shrinking code is exercised only once during a 16-minute run involving 500 concurrent GCs, a 45 GiB heap, and a 28 GiB young-gen size.  The change reduces the degenerated GCs from 6 to 5.

One reason that the remaining 5 degenerated GCs are not addressed by this change is that further work is required to handle a situation in which a requested TLAB is smaller than the available young-gen memory, but available memory is set aside in the evacuation reserve so cannot be provided to a mutator.  Future work will address this condition.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Change must be properly reviewed (1 review required, with at least 1 [Committer](https://openjdk.org/bylaws#committer))

### Reviewers
 * [Y. Srinivas Ramakrishna](https://openjdk.org/census#ysr) (@ysramakrishna - Author) ⚠️ Review applies to [774e07a1](https://git.openjdk.org/shenandoah/pull/180/files/774e07a19060f175db20295102ae4804818e61ed)
 * [William Kemper](https://openjdk.org/census#wkemper) (@earthling-amzn - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/shenandoah pull/180/head:pull/180` \
`$ git checkout pull/180`

Update a local copy of the PR: \
`$ git checkout pull/180` \
`$ git pull https://git.openjdk.org/shenandoah pull/180/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 180`

View PR using the GUI difftool: \
`$ git pr show -t 180`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/shenandoah/pull/180.diff">https://git.openjdk.org/shenandoah/pull/180.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp [hunk 1]
```diff
@@ -1225,120 +1222,152 @@ HeapWord* ShenandoahHeap::allocate_memory(ShenandoahAllocRequest& req, bool is_p
...
+  // reacquires the lock.  If that happens, we will need another recursive call to further reduce the size of our request
+  // for each time another thread allocates young memory during the brief intervals that the heap lock is available to
+  // interfering threads.  We expect this interference to be rare.  The recursion bottoms out when young_available is
+  // smaller than req.min_size().
+  ShenandoahAllocRequest smaller_req = ShenandoahAllocRequest::for_tlab(req.min_size(), smaller_lab_size);
+
+  // Note that shrinking the preferred size gets us past the gatekeeper that checks whether there's available memory to
+  // satisfy the allocation request.  The reality is the actual TLAB size is likely to be even smaller, because it will
+  // depend on how much memory is available within mutator regions that are not yet fully used.
+  HeapWord* result = allocate_memory_under_lock(smaller_req, in_new_region, is_promotion);
```
[the remark this answers is not in the document]
Comment (kdnilsen) [reply src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp:1385]: Thanks for your review.  I'm adding a comment to clarify the recursive algorithm and use of secondary ShenandoahAllocationRequest argument.

[the remark this answers is not in the document]
Comment (ysramakrishna) [reply src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp:1385]: ok, thanks!

**Review by earthling-amzn — APPROVED** [review]
Looks good. Thank you.

**Pull request closed without merging**